### PR TITLE
Lamb2

### PR DESCRIPTION
**Pull request by xdgarrido**
Oops, I did in the Google main. Lamb2 is fine to merge into the RocmPlatform/bert. 
Best regards,

Diego

### CODE REVIEW

**Comment by c0redumb** [comment]
Great. Thanks a lot.



Sent from my mobile. I do have fast fingers. Really.



-------- Original message --------

File: optimization.py [hunk 1]
```diff
@@ -163,7 +273,25 @@ def apply_gradients(self, grads_and_vars, global_step=None, name=None):
...
+      ############## BELOW ARE THE SPECIFIC PARTS FOR LAMB ##############
+
+      # Note: Here are two choices for scaling function \phi(z)
+      # minmax:   \phi(z) = min(max(z, \gamma_l), \gamma_u)
+      # identity: \phi(z) = z
+      # The authors does not mention what is \gamma_l and \gamma_u
+      # UPDATE: after asking authors, they provide me the code below.
+      # ratio = array_ops.where(math_ops.greater(w_norm, 0), array_ops.where(
+      #      math_ops.greater(g_norm, 0), (w_norm / g_norm), 1.0), 1.0)
+
```
Comment (micmelesse) [inline optimization.py:285]: I don't think the comments about the authors and the update is necessary as long the information is there but upto you.

File: optimization.py [hunk 2]
```diff
@@ -144,6 +154,106 @@ def apply_gradients(self, grads_and_vars, global_step=None, name=None):
           trainable=False,
           initializer=tf.zeros_initializer())
 
+      # Standard Adam update.
```
[the remark this answers is not in the document]
Comment (c0redumb) [reply optimization.py:157]: Diff is matching the second half of LAMB as a second half of ADAM. So "diff" indicates that there is a single block added in the middle, but indeed it is a single block added at the end. I think it is a none issue. "Diff" is not intelligent enough to tell exactly what you have added.

**Comment by c0redumb** [comment]
The default optimizer is "adam" if left unspecified. So we may not want to do "lamb" in the sample shell scripts.

**Comment by MikalaiDrabovich** [comment]
lgtm

**Comment by cagery** [comment]
Yes default is adam, for the completeness+convenience I'd still think the example should have that configurable knob in the example. As a test user, I pulled in the code and wanted to run the LAMB with example provided but the knob was missing.

**Comment by c0redumb** [comment]
Good question. This goes to the purpose of these test scripts. Indeed, the testers are just here to show that the model runs on **ROCm platform**. Runs = working and actually do training, not crashing, and uses multiple GPUs as intended. All the parameters do not matter (and they should not matter in 200 steps). This is not a script to measure performances, or to show people how to use the BERT model.

**Comment by cagery** [comment]
Okay I get your point, this is a tiny issue anyways. There is one more typo I think, in the readme file. Again tiny issue. Other than that I am ok. It works as I tried. Thanks.

**Pull request merged**